src/sequencer.c

```diff
@@ -0,0 +1,227 @@
+if sequence="59320,,123" then this is an absolute tick in the future and played only once
+you can remove scheduled sequences with sequencer=",,123" -- nothing in the tick section will delete a message if there is a tag
+tick takes precedence over divider if both given
+
+ok, how we deal with it internally:
+
+in the parse_amy_message, we check for sequence, if it's there, we set the event type to EVENT_SEQUENCE and send it to
+sequence_add_event(struct event e, tick, divider, tag) 
+
+sequence_add_event checks / parses the sequence string and fills in a new entry in the seqeunce entry LL
+it also can remove sequence entries if no tick is given and the tag exists 
+
+a thread/timer that checks every 48ppq tick with BPM 
+and then adds it when it's time to play
+it adds it by amy_adding_event with a re-set time variable and status = EVENT_SCHEDULED
+entries that have ticks get deleted after adding
+
+we also call a callback function if set void sequencer_callback(uint32_t tick)
+
+*/
+
+#include "sequencer.h"
+
+typedef struct  {
+    struct event e; // parsed event -- we clear out sequence and time here obv
+    uint32_t tag;
+    uint32_t tick; // 0 means not used 
+    uint16_t divider; // 0 means not used 
+} sequence_entry;
+
+// linked list of sequence_entries
+typedef struct sequence_entry_ll_t{
+    sequence_entry *sequence;
+    struct sequence_entry_ll_t *next;
+} sequence_entry_ll_t;
+
+sequence_entry_ll_t * sequence_entry_ll_start;
+
+// Optional sequencer hook that's called every tick
+void (*amy_external_sequencer_hook)(uint32_t) = NULL;
+
+// Our internal accounting
+uint32_t sequencer_tick_count = 0;
+uint64_t next_amy_tick_us = 0;
+uint32_t us_per_tick = 0;
+
+#ifdef ESP_PLATFORM
+#include "esp_timer.h"
+esp_timer_handle_t periodic_timer;
+#else
+#include <pthread.h>
+#endif
+
+void sequencer_recompute() {
+    us_per_tick = (uint32_t) (1000000.0 / ((amy_global.tempo/60.0) * (float)AMY_SEQUENCER_PPQ));    
+    next_amy_tick_us = amy_sysclock()*1000 + us_per_tick;
+}
+
+extern int parse_list_uint32_t(char *message, uint32_t *vals, int max_num_vals, uint32_t skipped_val);
+
+// Parses the tick, divider and tag out of a sequence message
+void parse_tick_and_tag(char * message, uint32_t *tick, uint16_t *divider, uint32_t *tag) {
+    uint32_t vals[3];
+    parse_list_uint32_t(message, vals, 3, 0);
+    *tick = (uint32_t)vals[0]; *divider = (uint16_t)vals[1]; *tag = (uint32_t)vals[2];
+}
+
+uint8_t sequencer_add_event(struct event e, uint32_t tick, uint16_t divider, uint32_t tag) {
+    // add this event to the list of sequencer events in the LL
+    // if the tag already exists - if there's tick/divider, overwrite, if there's no tick / divider, we should remove the entry
+    if(divider != 0 && tick != 0) { divider = 0; } // only use tick if both set
+
+    sequence_entry_ll_t *entry_ll = sequence_entry_ll_start;
+    sequence_entry_ll_t *prev_entry_ll = NULL;
```

Comment:
Isn't the normal way to do this as parent_p = &root_ptr, then access current as (*parent_ptr)...